frame/contracts/src/benchmarking/mod.rs

```diff
@@ -1583,6 +1587,786 @@ benchmarks! {
+				Instruction::Call(2), // call aux
+			])),
+			inject_stack_metering: true,
+			.. Default::default()
+		}));
+	}: {
+		sbox.invoke();
+	}
+
+	// w_call_indrect = w_bench - 3 * w_param
+	instr_call_indirect {
+		let r in 0 .. INSTR_BENCHMARK_BATCHES;
+		let num_elements = Contracts::<T>::current_schedule().limits.table_size;
+		use self::code::TableSegment;
+		use body::DynInstr::{RandomI32, Regular};
+		let mut sbox = Sandbox::from(&WasmModule::<T>::from(ModuleDefinition {
+			// We need to make use of the stack here in order to trigger stack height
+			// instrumentation.
+			aux_body: Some(body::plain(vec![
+				Instruction::I64Const(42),
+				Instruction::Drop,
+				Instruction::End,
+			])),
+			call_body: Some(body::repeated_dyn(r * INSTR_BENCHMARK_BATCH_SIZE, vec![
+				RandomI32(0, num_elements as i32),
+				Regular(Instruction::CallIndirect(0, 0)), // we only have one sig: 0
+			])),
+			inject_stack_metering: true,
+			table: Some(TableSegment {
+				num_elements,
+				function_index: 2, // aux
+			}),
+			.. Default::default()
+		}));
+	}: {
+		sbox.invoke();
+	}
+
+	// w_instr_call_indirect_per_param = w_bench - 1 * w_param
+	// Calling a function indirectly causes it to go through a thunk function whose runtime
+	// linearly depend on the amount of parameters to this function.
+	// Please note that this is not necessary with a direct call.
+	instr_call_indirect_per_param {
+		let p in 0 .. Contracts::<T>::current_schedule().limits.parameters;
+		let num_elements = Contracts::<T>::current_schedule().limits.table_size;
+		use self::code::TableSegment;
+		use body::DynInstr::{RandomI32, RandomI64Repeated, Regular};
+		let mut sbox = Sandbox::from(&WasmModule::<T>::from(ModuleDefinition {
+			// We need to make use of the stack here in order to trigger stack height
+			// instrumentation.
+			aux_body: Some(body::plain(vec![
+				Instruction::I64Const(42),
+				Instruction::Drop,
+				Instruction::End,
+			])),
+			aux_arg_num: p,
+			call_body: Some(body::repeated_dyn(INSTR_BENCHMARK_BATCH_SIZE, vec![
+				RandomI64Repeated(p as usize),
+				RandomI32(0, num_elements as i32),
+				Regular(Instruction::CallIndirect(p.min(1), 0)), // aux signature: 1 or 0
+			])),
+			inject_stack_metering: true,
+			table: Some(TableSegment {
+				num_elements,
+				function_index: 2, // aux
+			}),
+			.. Default::default()
+		}));
+	}: {
+		sbox.invoke();
+	}
+
+	// w_local_get = w_bench - 1 * w_param
+	instr_local_get {
+		let r in 0 .. INSTR_BENCHMARK_BATCHES;
+		use body::DynInstr::{RandomGetLocal, Regular};
+		let max_locals = Contracts::<T>::current_schedule().limits.stack_height;
+		let mut call_body = body::repeated_dyn(r * INSTR_BENCHMARK_BATCH_SIZE, vec![
+			RandomGetLocal(0, max_locals),
+			Regular(Instruction::Drop),
+		]);
+		body::inject_locals(&mut call_body, max_locals);
+		let mut sbox = Sandbox::from(&WasmModule::<T>::from(ModuleDefinition {
+			call_body: Some(call_body),
+			.. Default::default()
+		}));
```

Comment:
this feels quite repetitive

can we make a function that creates `Sandbox` from some code?

Reply:
Only the three benchmarks for the local instructions would profit from such a function. Every other benchmark needs some slight variation (setting some memory, globals, etc.). I don't think that this warrants a new function.

Reply:
You could use `sandbox(call_body).with_mem()` builder pattern in other cases, but it's up to you :)